pkg/controller/volume/persistentvolume/pv_controller.go

```diff
@@ -577,9 +577,10 @@ func (ctrl *PersistentVolumeController) syncVolume(volume *v1.PersistentVolume)
 		if err != nil {
 			return err
 		}
-		if !found && metav1.HasAnnotation(volume.ObjectMeta, pvutil.AnnBoundByController) {
```

Comment:
Can we add a unit test for this?

Reply:
I can try.

What exactly was the situation with external binder? That variant also seems to lack a unit test, or at least https://github.com/kubernetes/kubernetes/pull/67062/files didn't add one.

Reply:
I've added more tests and modernized the testing a bit. See individual commits for details.

I'm not sure whether the situation with the external binder is covered, but the code no longer checks for that, so I'd say it's not needed.

I've not found a way to store the PVC in the informer cache, so the code now falls back all the way to a Get operation that then gets satisfied by the reactor.

I've verified manually that this happens and added a corresponding test case where the PVC really doesn't exist. I briefly checked whether it would be possible to validate that the Get happens resp. doesn't happen, but found no hook in the existing framework for such an extra check and would prefer to not make changes that affect other tests.

I also need to get back to some other work :sweat_smile: